source/common/router/scoped_rds.cc

```diff
@@ -308,6 +308,22 @@ void ScopedRdsConfigSubscription::onRdsConfigUpdate(const std::string& scope_nam
 void ScopedRdsConfigSubscription::onConfigUpdate(
     const std::vector<Envoy::Config::DecodedResourceRef>& resources,
     const std::string& version_info) {
+  Protobuf::RepeatedPtrField<std::string> to_remove_repeated;
+  for (const auto& scoped_route : scoped_route_map_) {
+    *to_remove_repeated.Add() = scoped_route.first;
+  }
+  onConfigUpdate(resources, to_remove_repeated, version_info);
+}
+
+Protobuf::RepeatedPtrField<std::string>
+ScopedRdsConfigSubscription::detectUpdateConflictAndCleanupRemoved(
```

Comment:
the order is problematic here:  
lets say we have {h1:scope1, h2:scope2}, and the update is like {scope1:h2, scope2:h3}
the update will be rejected as the hash of new scope1 is conflicting the original scope2. 
But you can see we should not reject that, this is not introduced by your change, but rather something we can improve, could you make this improvement? 

also remember deletions are done ahead of add/updates, let's add a final-state hash, apply deletions first, apply ALL the add/updates, and then count the hash conflict.

Reply:
The update won't be rejected in this case.
There are 3 cases where update will be rejected:
In sotw api, when the incoming resources have key or scope name conflict.
In delta api, when the incoming resources have key or scope name conflict.
In delta api, when the incoming resources have key or scope name conflict with old scope that are not removed.

Reply:
In the new procedure, in an sotw update, we mark all old scopes as removed scopes from the beginning, and conflict key and scope name with removed scopes will be ignored.

Reply:
I am not following.
I think SotW and Delta form now uses the same verification logic. 

There might be some confusion, but the case I was proposing is like "an incoming resource has a hash conflict with an existing scope, which will be updated by the same update but haven't yet".

Reply:
right, in Delta APi, this is still a probelm. I missed your last reply, sorry.